### PR TITLE
Remove cylc.profiling package and cylc-profile-battery

### LINKED ISSUE #2991: cylc command error when running in Anaconda or virtualenv (or both)
When Cylc is packaged with `setuptools`, and we execute the `cylc` entry command, it may get confused and fail, depending on the values of the `$PATH` variable.

In my case, I have copied `./usr/bin/cylc` to `~/bin/cylc`. And if that script is the first match for `cylc`, then everything works.

Now, after enabling Anaconda Python 3, creating a virtualenv, and installing with the code from #2990 , running `cylc version --long` fails with:

```
fatal: not a git repository: '/home/kinow/Development/python/workspace/cylc/venv/.git'
No version is set, cylc must be built first. Aborting...
```

The issue is that the function to find Cylc's version uses the location of the `cylc` utility to guess where is Cylc's home. Then tries to open the `git` folder to confirm it exists, and finally executes the git command to retrieve the version.

But when we create the scripts via `setuptools`, `cylc` gets installed in the user PATH automatically. And in the case where the user has Anaconda Python, it gets installed in some directory within Anaconda installation. Or if using virtualenv, it might be within the `venv` folder, added to the `$PATH` automagically too.

This is breaking the Travis-CI tests in the `setup.py` pull request, and caused the unit tests to fail in my environment for a while until I realized what was going on.

### PR DESCRIPTION
**Pull request by kinow**
For #2991 (once this one gets merged, will test and close #2991 if no other command failed in my Anaconda environment).

The `cylc-profile-battery` command tries to locate the directory of `cylc` script, and locate the base Cylc installation directory. Then it fails if that directory is not a Git repository (i.e. does not have a `.git` folder).

This causes issues when running the tests with Cylc being installed via `pip` under a virtual environment in Anaconda.

~In #2997 we changed the way the Cylc version is located, so that there is no more need to have a git directory. And now in this pull request we are using that approach instead of using the `git` command line.~

~And instead of trying to locate the Cylc directory, the script will default to the old behaviour of assuming `.` is a Cylc installation of `.git`, but also accept an optional parameter that defines where is the Cylc directory we would like to use in the experiments.~

As per discussion (see comments below) this PR removes the `cylc.profiling` package, the `cylc-profile'-battery` and updates the documentation.

### CODE REVIEW

**Comment by kinow** [comment]
First executed `cylc profile-battery -e hello-world` and got the experiment running successfully.

```
...
$ rm -rf /home/kinow/cylc-run/profile-15524351935191412

Version  Run  Elapsed Time (s)  CPU Time - Total (s)  Max Memory (kb)  
HEAD     run  2.97              0.6736363636363637    35721.09090909091

Results for experiment "hello-world" have been written out to "/home/kinow/Development/python/workspace/cylc/.profiling/plots/hello-world-1552435229"

```

And after the change it works too after removing my `.profiling` directory: 

```
...
$ rm -rf /home/kinow/cylc-run/profile-15524366025308645

Version  Run  Elapsed Time (s)    CPU Time - Total (s)  Max Memory (kb)
HEAD     run  2.9518181818181817  0.6736363636363635    35728.0        

Results for experiment "hello-world" have been written out to "/home/kinow/Development/python/workspace/cylc/.profiling/plots/hello-world-1552436637"
```

Then I did the following test.

- `kinow@/home/kinow: rm -rf .profiling`
- `kinow@/home/kinow: mv .git .git2`
- `kinow@/home/kinow: cd /tmp && git clone git clone https://github.com/cylc/cylc.git && cd -`
- `cylc profile-battery -e hello-world`: Fails with _ERROR: profiling requires cylc to be a git repository._
- `cylc profile-battery -e hello-world /tmp/cylc`

The experiment is executed successfully, and the `.profiling` folder was correctly created under `/tmp/cylc/.profiling/`:

```
...
$ rm -rf /home/kinow/cylc-run/profile-15524424398637655

Version  Run  Elapsed Time (s)  CPU Time - Total (s)  Max Memory (kb)   
HEAD     run  2.89090909090909  0.6309090909090909    35690.545454545456

Results for experiment "hello-world" have been written out to "/tmp/cylc/.profiling/plots/hello-world-1552442474"
```

**Comment by kinow** [comment]
Added a couple `TODO` markers, as this current approach requires to change a constant in the `cylc.profiling` package - even though constants should be immutable.

To remove these `TODO` markers, I reckon we would have to modify the current design of the `cylc.profiling` package, which would take longer. Hopefully this current fix is sufficient to proceed with this issue and #2990 (blocked by this one ATM).

**Comment by hjoliver** [comment]
@kinow - I haven't looked at your code yet, but for use with installed versions (not git clones) we should probably assume our recommended side-by-side installation structure (e.g. `/usr/local/bin/cylc` uses `$CYLC_VERSION` to find `/opt/cylc/$CYLC_VERSION/`) ... possibly overridden by optional explicit locations as a quick read above suggests you've done.

**Comment by hjoliver** [comment]
Or (sorry - not thinking!) are we expecting use of `cylc profile-battery` only with git clones?

**Comment by kinow** [comment]
>Or (sorry - not thinking!) are we expecting use of cylc profile-battery only with git clones?

I think this is a requirement for `cylc profile-battery`. You must provide a folder with a clone of Cylc, so that it is able to either use the current version, or run the experiment against multiple versions and report the benchmarking results (that's my current high-level understanding of that package/tool).

**Comment by kinow** [comment]
When users install `cylc` in the future, possibly via `pip install cylc`... they will still get `cylc profile-battery`. But installing via `pip`, there won't be a `.git` repository anywhere.

So users will now be required to clone the repository somewhere, in order to use this tool. Not sure if the best approach, but at least this way `cylc profile-battery` can still be used even after Anaconda, pip, setup.py, etc.... (and I can continue testing our current work on the `setup.py` branch :grin:  ).

**Comment by hjoliver** [comment]
(p.s. no need to reply if you've left work already - I'll take a look at the code after I've finished agenda wrangling!)

**Comment by hjoliver** [comment]
(OK ... you already replied - I didn't reload the page!)

**Comment by hjoliver** [comment]
Makes sense; let's see what @oliver-sanders thinks.

**Comment by oliver-sanders** [comment]
> So users will now be required to clone the repository somewhere

Yes. The purpose of the `cylc profile-battery` is to allow developers to profile their development branches. It is only expected to work on a development clone. I can't think why we would want it to work on an installation via pip/conda.

**Comment by hjoliver** [comment]
> I can't think why we would want it to work on an installation via pip/conda.

I can imagine that site admins who install Cylc in critical production environments might want to check that performance has not degraded with the next release (possibly with a local suite).

We don't have to support that, of course (and maybe we shouldn't) ... but I imagine it would be even easier to compare to installed releases than two git versions checked out on the fly.

**Comment by hjoliver** [comment]
If we don't want to support that, can we have `cylc profile-battery` removed from releases?

**Comment by kinow** [comment]
>If we don't want to support that, can we have cylc profile-battery removed from releases?

Maybe we could promote it to a top level project? `cylc/profile-battery` or `cylc/cylc-profile-battery`. It could use `cylc` as a dependency, and be a standalone project.

Site admins could then manage where to install that, how to use, etc.

**Comment by hjoliver** [comment]
I quite like that idea.

**Comment by oliver-sanders** [comment]
> I quite like that idea.

Ditto!

Though we would need to be slightly careful not to break it.

The `cylc profile-battery` could really do with a re-write (part of the reason it's so horrible is that we needed lots of back-compatibility for a report). I got part way through and had to bail to get other work done.

**Comment by kinow** [comment]
So what about this plan of action?

1. Use this fix so that the work on the `setup.py` can continue for now
2. I will create a separate ticket to move `cylc-profile-battery` to its own repository
3. I will mark that ticket as BLOCKED by the setup.py ticket (i.e. work on the `cylc-profile-battery` depends first on being able to use something like `from cylc import __version__` ...)

**Comment by kinow** [comment]
And I really liked the code of the profile battery. But if you think it could be improved, then I really want to see the new design and play with it! I suspect having it in its own repository will make it actually easier to re-write it.

**Comment by hjoliver** [comment]
@kinow 's plan of action looks good to me.  I'll defer to @oliver-sanders as the code owner though.  (And perhaps @oliver-sanders can also check that his UK colleagues don't object).

**Comment by oliver-sanders** [comment]
Possible alternative:

1. Delete `cylc profile-battery`.
2. Raise an issue to move it.

We've already deleted `cylc review` with plans to re-incarnate in the near future so that wouldn't be too bad.

**Comment by kinow** [comment]
Oh, that would be even easier! If others are OK with that too, I will update (or close an raise a new one) the pull request. Cheers!

**Comment by matthewrmshin** [comment]
@kinow please go ahead.

**Comment by kinow** [comment]
Pull request update to now remove the cylc.profiling package, the cylc-profile-battery, and update documentation.

All tests passing, and my ol' suite number five ran without issues locally.

Ready for review again.

Cheers
Bruno

File: etc/cylc-bash-completion [hunk 1]
```diff
@@ -38,7 +38,7 @@ _cylc() {
...
     opts="$(cylc print -x -y 2>/dev/null)"
-    suite_cmds="broadcast|bcast|cat-log|log|cat-state|check-versions|checkpoint|diff|compare|documentation|browse|dump|edit|ext-trigger|external-trigger|get-directory|get-suite-config|get-config|get-suite-version|get-cylc-version|graph|graph-diff|hold|import-examples|insert|jobscript|kill|list|ls|ls-checkpoints|monitor|nudge|ping|poll|print|profile-battery|register|release|unhold|reload|remove|report-timings|reset|restart|run|start|scan|scp-transfer|search|grep|set-verbosity|show|spawn|stop|shutdown|submit|single|suite-state|test-battery|trigger|validate|view|warrenty"
+    suite_cmds="broadcast|bcast|cat-log|log|cat-state|check-versions|checkpoint|diff|compare|documentation|browse|dump|edit|ext-trigger|external-trigger|get-directory|get-suite-config|get-config|get-suite-version|get-cylc-version|graph|graph-diff|hold|import-examples|insert|jobscript|kill|list|ls|ls-checkpoints|monitor|nudge|ping|poll|print|register|release|unhold|reload|remove|report-timings|reset|restart|run|start|scan|scp-transfer|search|grep|set-verbosity|show|spawn|stop|shutdown|submit|single|suite-state|test-battery|trigger|validate|view|warrenty"
```
Comment (hjoliver) [inline etc/cylc-bash-completion]: Not your fault @kinow but I see a typo at the end of this line!  ("warr**e**nty" should be "warr**a**nty")

Comment (kinow) [reply etc/cylc-bash-completion]: I had a few typos that I wanted to get fixed too @hjoliver , so prepared a pull request to address them :+1: 

See #3013 

Thanks!

**Comment by matthewrmshin** [comment]
This is now in conflict now that the typos branch is merged.

**Comment by kinow** [comment]
>This is now in conflict now that the typos branch is merged.

Thanks @matthewrmshin ! I excluded `cylc-profile-battery` and another file because I thought it would cause conflicts, but forgot about `etc/cylc-bash-completion`.

Conflict undone!

**Comment by kinow** [comment]
>In fact I think all the dev suites are performance examples so we could probably move them all?

Oh, I had no idea. Just pushed another commit that removes `etc/dev-suites`. If those suites were not supposed to be removed, let me know and I will drop the commit :+1: Thanks!!!

**Comment by oliver-sanders** [comment]
I just spotted there are some `xtrigger` suites in there, not too sure why they are in `dev-suites`, should they be moved to `examples` @hjoliver?

**Comment by hjoliver** [comment]
1.  `etc/dev-suites/` is a historical dumping ground for suites used for manually testing or illustrating some kind of performance issue
1. it also holds suites used as built-in cases for `cylc profile-battery` (some of these originate from 1.)
1. the xtrigger examples should be moved to `/etc/examples/`, I think I just forgot to do that when finishing up the xtrigger development last year.

**Comment by hjoliver** [comment]
Just wondering about the logic of moving profiling to a new repository vs leaving the test battery in the main repository.

I guess that makes sense because the test battery is to check that the current source version passes all tests, but the profiler compares two arbitrary cylc versions - i.e. it is not inherently tied to the current source version.

**Comment by hjoliver** [comment]
In principle the test battery could also be used in a release, after installation, to check that everything works.

Any other developer-specific files - these days just a few files under `etc/` should probably be removed from main repo.

**Comment by kinow** [comment]
Good point @hjoliver , I hadn't considered the test-battery. I think it makes sense to move the profile-battery as it can be maintained as a separate project, with its own release cycle, bugs/issues/pull requests/features.

And the profile-battery should not change if we change Cylc's internal API. So there is a more clear separation between Cylc and profile-battery. It could be possible that it would still work for Cylc 9 or Cylc 10, maybe with some minor adjustments.

But the test-battery relies in some parts on API in Cylc that is not exposed to public. Some tests verify - as far as I could tell - if parsing is working, task pool state, messages, etc. So this one is much more tangled with Cylc.

Changing features in Cylc require changing test-battery some times. But not the profile-battery I think (maybe create a new experiment?).

**Comment by hjoliver** [comment]
Yes, that's right. So we're doing the right thing - profiler out, test battery in.

**Comment by kinow** [comment]
>the xtrigger examples should be moved to /etc/examples/, I think I just forgot to do that when finishing up the xtrigger development last year.

Done in 8f2b749745d0aaa965c1cbf06287e1fba5468a71! :tada:

**Pull request merged**